### PR TITLE
Resilient connection

### PR DESCRIPTION
**Pull request by chkeita**
## Summary of the Pull Request

Using an exponential backoff strategy for the http calls in the agent.

## PR Checklist
* [x] Applies to work item: #79 
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/onefuzz) and sign the CLI.
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

_What does this include?_

## Validation Steps Performed

_How does someone test & validate?_

### CODE REVIEW

File: src/agent/utils/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            let response = cloned
+                .send()
+                .await
+                .map_err(|err| backoff::Error::Permanent(anyhow::Error::from(err)))?
+                .error_for_status()
+                .map_err(|err| {
+                    log::warn!("Transient error {}", err);
+                    backoff::Error::Transient(anyhow::Error::from(err))
+                })?;
+            Result::<Response, backoff::Error<anyhow::Error>>::Ok(response)
```
Comment (chkeita) [inline src/agent/utils/src/lib.rs:48]: The type inference engine needed a little help here

File: src/agent/utils/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,60 @@
+use anyhow::Result;
```
Comment (bmc-msft) [inline src/agent/utils/src/lib.rs:1]: Needs copyright statement

File: src/agent/Cargo.toml [hunk 3]
```diff
@@ -7,6 +7,7 @@ members = [
     "onefuzz-agent",
     "onefuzz-supervisor",
     "storage-queue",
+    "utils",
```
Comment (bmc-msft) [inline src/agent/Cargo.toml:10]: Can this go in a `retry` module of the `onefuzz` crate?

File: src/agent/utils/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+// Licensed under the MIT License.
+
+use anyhow::Result;
+use async_trait::async_trait;
+use backoff::{self, future::FutureOperation, ExponentialBackoff};
+use reqwest::Response;
+use std::time::Duration;
+
+const DEFAULT_RETRY_PERIOD: Duration = Duration::from_millis(500);
+const MAX_ELAPSED_TIME: Duration = Duration::from_secs(60 * 5);
```
Comment (bmc-msft) [inline src/agent/utils/src/lib.rs:11]: While reviewing this, I found [Azure retry guidance](https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#general-rest-and-retry-guidelines).

We probably want to tune this to align with the conservative side of Azure retry guidance.

* Max of 5 total tries (4 retries plus the initial)
* Max elapsed duration of 30s
* Support error code specific knowledge to enable retries (only retry if it's a connection issue OR one of these HTTP error codes)
    * 408 Request Timeout
    * 429 Too Many Requests
    * 500 Internal Server Error 
    * 502 Bad Gateway
    * 503 Service Unavailable
    * 504 Gateway Timeout

**Review by bmc-msft — COMMENTED** [review]
In general, this is extremely positive change.  

I didn't know defining traits for third-party structs was allowed.  This is significantly better than the exponential backoff code added to [AVML](https://github.com/microsoft/avml/blob/4409f0048854e44d9b4d0f2c31261acb43174e92/src/blobstore.rs#L97)

File: src/agent/Cargo.toml (hunk 3 above)
Comment (chkeita) [reply src/agent/Cargo.toml:10]: I will have to reference onfuzz in the storage-queue and proxy-managed crates

Comment (bmc-msft) [reply src/agent/Cargo.toml:10]: can it be named something other than utils then?  we have a large number of `utils` spread across multiple crates as is.

**Review by bmc-msft — APPROVED** [review]

**Comment by bmc-msft** [comment]
Every task in testing is failing with

```
[
'task failed. exit_status:code=1 signal=None success=False', 
'', 
"[2020-10-26T22:43:50Z ERROR onefuzz_agent] error running task: Couldn't upload file.  path:/onefuzz/3f059c05-2b1b-44b9-985f-ae1c6135ad1b/task_inputs_2/eb408ddc4fa484e6befdf5954e56a2198c7a9fab dir:task_inputs_2 err:This request cannot be retried because it cannot be cloned\nError: Couldn't upload file.  path:/onefuzz/3f059c05-2b1b-44b9-985f-ae1c6135ad1b/task_inputs_2/eb408ddc4fa484e6befdf5954e56a2198c7a9fab dir:task_inputs_2 err:This request cannot be retried because it cannot be cloned\n"
]
```

File: src/agent/reqwest-retry/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+const DEFAULT_RETRY_PERIOD: Duration = Duration::from_secs(2);
+const MAX_ELAPSED_TIME: Duration = Duration::from_secs(30);
+const MAX_RETRY_ATTEMPTS: i32 = 5;
+
+fn to_backoff_response(
+    result: Result<Response, reqwest::Error>,
+) -> Result<Response, backoff::Error<anyhow::Error>> {
+    match result {
+        Err(error) => Err(backoff::Error::Permanent(anyhow::Error::from(error))),
```
Comment (chkeita) [inline src/agent/reqwest-retry/src/lib.rs:21]: Need more handling here for socket level error like this one 
```
{
	"code": 468,
	"errors": ["task failed. exit_status:code=1 signal=None success=False", "", "[2020-10-26T23:04:11Z ERROR onefuzz_agent] error running task: Couldn't upload file.  path:/onefuzz/23fd2e08-df2d-44b5-8af6-d87013497732/task_inputs_2/d1da9c2db5e6e9903cc2d9bd1235d0a1ddf8f48b dir:task_inputs_2 err:error sending request for url (https://fuzze3j4evu5lzzfo.blob.core.windows.net/oft-inputs-83388a5a2f985661b81bf7b96c1554d2/d1da9c2db5e6e9903cc2d9bd1235d0a1ddf8f48b?<sas token>): connection error: Connection reset by peer (os error 104)\nError: Couldn't upload file.  path:/onefuzz/23fd2e08-df2d-44b5-8af6-d87013497732/task_inputs_2/d1da9c2db5e6e9903cc2d9bd1235d0a1ddf8f48b dir:task_inputs_2 err:error sending request for url (https://fuzze3j4evu5lzzfo.blob.core.windows.net/oft-inputs-83388a5a2f985661b81bf7b96c1554d2/d1da9c2db5e6e9903cc2d9bd1235d0a1ddf8f48b?<sas toke>): connection error: Connection reset by peer (os error 104)\n"]
}
```

**Review by bmc-msft — APPROVED** [review]

**Comment by bmc-msft** [comment]
Performing integration testing now

**Pull request merged**